### PR TITLE
Lula San - Phoenix

### PR DESCRIPTION
**Pull request by machilula**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,40 @@
...
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.shape(
+    { time: PropTypes.string.isRequired }
+  ).isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLikeEntry: PropTypes.func.isRequired
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:37]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did.

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.shape(
+    { time: PropTypes.string.isRequired }
+  ).isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:35]: 👀 Be sure to keep an eye out for warnings about the actual _type_ of any props. These `propTypes` describe `timeStamp` as an object with inner structure, when in fact the data available from the messages is just a string. The `TimeStamp` component does expect to receive the `time` prop as a string, but the lowercased `timeStamp` here is just the name of a prop and is not the same as the uppercase `TimeStamp` name of the component that `ChatEntry` is using internally, so doesn't have any nested object meaning. The names we use to refer to a value as it travels down through a series of props can change with each encountered component since they are named in a way that makes local sense for that component, which can sometimes make a prop name resemble other values or components used in the system. But we need to be careful to keep their meanings distinct in our heads.

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,40 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({id, sender, body, timeStamp, liked, onLikeEntry }) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params) I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 4]
```diff
@@ -1,20 +1,40 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={`chat-entry ${isLocal}`}>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:18]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.shape(
+        { time: PropTypes.string.isRequired }),
+      liked: PropTypes.bool.isRequired
+    })
+  ).isRequired,
+  onLikeEntry: PropTypes.func.isRequired
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:33]: Similar to the props for `ChatEntry` here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+ChatLog.propTypes = {
+  entries: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.shape(
+        { time: PropTypes.string.isRequired }),
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:29]: As with `ChatEntry`, this should be just a string rather than a shape.

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry.jsx';
+
+
+const ChatLog = ({entries, onLikeEntry}) => {
+  const entryComponents = entries.map((entry) => {
+    return (
+      <ChatEntry
+        key={entry.id}
+        id={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:11]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import './ChatLog.css';
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry.jsx';
+
+
+const ChatLog = ({entries, onLikeEntry}) => {
+  const entryComponents = entries.map((entry) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:7]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/components/ChatLog.jsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        key={entry.id}
+        id={entry.id}
+        sender={entry.sender}
+        body={entry.body}
+        timeStamp={entry.timeStamp}
+        liked={entry.liked}
+        onLikeEntry={onLikeEntry}
+      />);
+  });
+  return (<ul className="chat-log">{entryComponents}</ul>);
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:19]: Since you're using an `ul` tag to list your chat entries, the child of this should be `li` tags. This can be accomplished either by having `ChatEntry` use `li` as its top-level tag (yours uses `div`) or by wrapping each `ChatEntry` in an `li` in the mapping step. We should aim to generate valid HTML once all the JSX components have been expanded to actual tags.

Comment (anselrognlie) [reply src/components/ChatLog.jsx:19]: Nit: this `return` easily fits on one line, so no need to wrap the expression in `()`.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,45 @@
 import './App.css';
+import ChatLog from './components/ChatLog';
+import { useState } from 'react';
+import messageData from './data/messages.json';
 
 const App = () => {
+  const [entryData, setEntryData] = useState(messageData);
+
+  const handleLikeEntry = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:9]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,45 @@
...
+    setEntryData(entryData => entryData.map(entry =>{
```
Comment (anselrognlie) [inline src/App.jsx:10]: Nice use of the callback setter style. In this application, it doesn't really matter whether we use the callback style or the value style, but it's good practice to get in the habit of using the callback style.

Comment (anselrognlie) [reply src/App.jsx:10]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 12]
```diff
@@ -1,14 +1,45 @@
...
+      if (entry.id === id) {
+        return {...entry, liked: !entry.liked};
```
Comment (anselrognlie) [inline src/App.jsx:12]: We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 13]
```diff
@@ -1,14 +1,45 @@
...
+    let total = 0;
+    for (const entry of entryData) {
+      if (entry.liked) {
+        total += 1;
+      }
+    };
+    return total;
+  };
+
+  const totalLikes = calculateTotalLikes(entryData);
```
Comment (anselrognlie) [inline src/App.jsx:29]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Explicitly totalling the count is perfectly fine, but many JS programmers would use `reduce` to achieve this.
```js
  const calculateTotalLikes = (entryData) => {
    return entryData.reduce((acc, entry) => {
      return entry.liked ? acc + 1 : acc;
    }, 0);
  };
```
The first few times we work with `reduce`, it can be challenging to understand, but it's a tool that gets used commonly enough that it's worth practicing.

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,40 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({id, sender, body, timeStamp, liked, onLikeEntry }) => {
+  const likeButton = liked ? '❤️' : '🤍';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 15]
```diff
@@ -1,20 +1,40 @@
...
+  const isLocal = sender === 'Vladimir' ? 'local' : 'remote';
+
+  const onLikeToggle = () => {
+    onLikeEntry(id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:10]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 16]
```diff
@@ -1,20 +1,40 @@
...
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
+        </p>
+        <button
+          className="like"
+          onClick={onLikeToggle}>{likeButton}
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:22]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

File: src/components/ChatEntry.jsx [hunk 17]
```diff
@@ -1,20 +1,40 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({id, sender, body, timeStamp, liked, onLikeEntry }) => {
+  const likeButton = liked ? '❤️' : '🤍';
+  const isLocal = sender === 'Vladimir' ? 'local' : 'remote';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: Nice logic to decide whether to treat a message as local or remote. How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatEntry` itself have enough information as it is to "know" which messages are local and which are remote?

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
